### PR TITLE
[DataGrid] Make autoHeight work

### PR DESCRIPTION
**Pull request by DanailH**
#604 

Ok so the PR is smaller than expected 😅

SO what I did - > I tried moving the `<AutoSizer>` component down to the `<GridWindow>` so basically 

```
<AutoSizer onResize={onResize} nonce={props.nonce}>
  <GridWindow ref={windowRef}>
    <Viewport ref={renderingZoneRef} />
  </GridWindow>
</AutoSizer>
```

The issue with that approach is that the `apiRef` then is messed up, meaning the trid is rendered before all the methods in the `apiRef` are loaded. The `AutoSizer` prevents this because it doesn't render the children if the container doesn't have dimensions (that is not part of the state of the grid, it comes from an internally kept state inside the `AutoSizer` itself). 

I tried replicating the same logic in the `GridRoot` component as that's the one that is the entry point after moving the `AutoSizer` down. It can be done with a little bit of abstraction and suffering logic around.

But then I tried to fix it by just changing the dimensions the `AutoSizer` adds and it seems to work. I tested it with the demos in the Storybook and in the Docs and it seems to be fine.

**I'll mark it as a draft so we can discuss if this approach works.**

The Draft is over. The solution here was to add a wrapper around the grid and keep the `AutoSizer` where it currently is (for now at least). Once we have solved the problem with the `apiRef` we can have the effort to move the AutoSizer and calculate the dimensions of the grid more reliably.

```
<GridContainer>
  <AutoSizer onResize={onResize} nonce={props.nonce}>
```

### CODE REVIEW

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
height 0 is important, it can't be removed. Did you plan the line?

**Comment by oliviertassinari** [comment]
> Did you plan the line?

We land on #350 but it's really noisy. If the fix was one line, it would have been the inverse of the change of this PR.

**Comment by oliviertassinari** [comment]
> The issue with that approach is that the apiRef then is messed up, meaning the trid is rendered before all the methods in the apiRef are loaded.

I had experienced the same the last time I had a quick look at the problem. IMHO, it's an issue with the `apiRef` and the state structure of the grid. I suspect the same underlying structure issue as with strict mode.

**Comment by dtassone** [comment]
Fix #604 

The total height of the grid component is stored in the state `state.containerSizes.totalSizes.height` so have you tried adding a div before Autosizer and apply the height in an effect...?

**Comment by oliviertassinari** [comment]
> The total height of the grid component is stored in the state state.containerSizes.totalSizes.height so have you tried adding a div before Autosizer and apply the height in an effect?

Isn't that hiding the root issue? Shouldn't reproduce the approach of ag-Grid?

**Comment by oliviertassinari** [comment]
Do we even need Autosize when autoHeight is set?

**Comment by DanailH** [comment]
> The total height of the grid component is stored in the state `state.containerSizes.totalSizes.height` so have you tried adding a div before Autosizer and apply the height in an effect?

Even if that works (I'll try it) it would be better to move the `AutoSizer` down because the part that has a "dynamic" height is the rows. The rest is static so it shouldn't be dependent on it. I can try and solve the issue with the missing methods in the `apiRef`.

@oliviertassinari thanks for sending the ticket and the PR I'll check them.

**Comment by dtassone** [comment]
> > The total height of the grid component is stored in the state `state.containerSizes.totalSizes.height` so have you tried adding a div before Autosizer and apply the height in an effect?
> 
> Even if that works (I'll try it) it would be better to move the `AutoSizer` down because the part that has a "dynamic" height is the rows. The rest is static so it shouldn't be dependent on it. I can try and solve the issue with the missing methods in the `apiRef`.
> 

Yes, it should be fine but it's a bigger refactoring. 
I think, anyhow, we will need to wrap autosizer with a div, and do a bit of a hack either like this workaround or by putting the height 🤔 
https://github.com/mui-org/material-ui-x/issues/604#issuecomment-756499141

**Comment by DanailH** [comment]
Ok I spend some time yesterday investigating more in-depth how the `apiRef` is being built. So it is created [HERE](https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/GridComponent.tsx#L60). Then we have the grid state initializer [HERE](https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/GridComponent.tsx#L61) and that one adds the basic state API. After that, the rest of the hooks are fired but they don't add their methods to the `apiRef` instance immediately, as the state does.

So basically what I'm saying is that line [HERE](https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/components/AutoSizer.tsx#L152) in the `AutoSizer` is the one that "saves" it from throwing errors because of missing APIs. 

On the initial render the `apiRef` only has the default state API and on every continent render it has all the rest APIs. That's why if I move the `AutoSizer` things start to break.

One solution will be to keep that check on the `GridComponent` lvl, or check if the `apiRef` has all the needed methods in it, which would be better, and then render the grid.

The last option and the most correct one would be to fix the problem at the source, a.k.a make sure that after each hook the API is correctly added to the `apiRef`, which is a different problem from the `autoHeight` issue.

**Comment by DanailH** [comment]
Ok, I've updated the solution and added a test case.

We had a call with @dtassone to discuss potential solutions and the `apiRef`. Since the strict mode issue will have to be fixed as well it would be better to do that as part of the strict mode PR, not this one. The solution here was to add a wrapper around the grid and keep the `AutoSizer` where it currently is (for now at least). Once we have solved the problem with the `apiRef` we can have the effort to move the `AutoSizer` and calculate the dimensions of the grid more reliably.

**Review by oliviertassinari — COMMENTED** [review]
Did you try to apply the same method by replacing the current AutoSizer with a NoSsr component (to delay by one effect) and move deeper in the DOM the AutoSizer?

**Comment by dtassone** [comment]
Ok I have made some progress with the strict mode, and I have 2 potential solutions for the issue with the missing method on initial start. 
1. Split the current hooks, in 2, one part for the api and one part to coordinate the component. 
Ie.  we split useRows, in 2 useRowsApi and useRows
We then call useSomethingApi in useApiRef, so the apiRef has all its methods.
2. We initialize the apiRef with a default GridApi implementing all members in useApiRef

**Review by dtassone — APPROVED** [review]

File: packages/grid/_modules_/grid/components/containers/GridContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  const gridWrapperRef = React.useRef<HTMLDivElement>(null);
+  React.useEffect(() => {
+    const gridWrapperDiv = gridWrapperRef.current;
+    if (gridWrapperDiv && gridOptions.autoHeight) {
+      const gridWrapperElement: HTMLDivElement = gridWrapperDiv.querySelector('div')!;
+      gridWrapperElement.style.height = '';
+    }
+  }, [gridWrapperRef, gridOptions]);
+
+  return gridOptions.autoHeight ? (
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/containers/GridContainer.tsx:21]: Why is it only an issue for auto height?

File: packages/grid/_modules_/grid/components/containers/GridContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+export function GridContainer(props: GridContainerProps) {
+  const apiRef = React.useContext(ApiContext);
+  const gridOptions = useGridSelector(apiRef, optionsSelector);
+
+  const gridWrapperRef = React.useRef<HTMLDivElement>(null);
+  React.useEffect(() => {
+    const gridWrapperDiv = gridWrapperRef.current;
+    if (gridWrapperDiv && gridOptions.autoHeight) {
+      const gridWrapperElement: HTMLDivElement = gridWrapperDiv.querySelector('div')!;
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/containers/GridContainer.tsx:16]: Are you getting the child of the wrapper here?

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
It's not working. I have taken the reproduction of #344 and use the package version. It grows forever: https://codesandbox.io/s/brave-chatelet-56z4o?file=/src/Demo.js. It's a regression.

It sounds like we need to add a test case in the CI for #344. Oops, I should have done it in my #350 PR 😇. It's still not too late, quite the opposite :).

**Comment by DanailH** [comment]
> Did you try to apply the same method by replacing the current AutoSizer with a NoSsr component (to delay by one effect) and move deeper in the DOM the AutoSizer?

No, I didn't, I wasn't aware that that is an option. We've been discussing fixing the `apiRef` not having a workaround. I've checked the other PR and continue working on it. After all, what is in that PR was what I wanted to do in the first place. If everyone is ok with using `NoSsr` and basically loading the grid on the second `useEffect` the other solution is better.

**Comment by DanailH** [comment]
I'm closing this PR in favour of https://github.com/mui-org/material-ui-x/pull/940

**Comment by oliviertassinari** [comment]
> If everyone is ok with using NoSsr and basically loading the grid on the second useEffect

True, I have used NoSsr to move faster in the POC, we don't necessarily have to use the component, we could unbundle it and make the useEffect sequence more explicit with a new custom component.

**Pull request closed without merging**